### PR TITLE
Implement ipv6-address-token-id key (LP: #1737976)

### PR DESCRIPTION
**Pull request by slyon**
## Description
This allows to statically configure the IPv6 host ID (low 64 bits) when auto-generated IPv6 addressing is used (i.e. DHCPv6 stateless, SLAAC).

It introduces one new YAML key in the schema: `ipv6-address-token-id`.

Fixes LP: #1737976

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented.
- [x] \(Optional\) Closes an open bug in Launchpad: https://bugs.launchpad.net/ubuntu/+source/nplan/+bug/1737976

### CODE REVIEW

**Comment by slyon** [comment]
Hey @vorlonofportland I'd like to get your opinion on the the schema key wording: `ipv6-address-token-id`
I wasn't really sure how to name it... what do you think about it?

**Comment by cluonbeam** [comment]
You could drop the `id` portion of the schema key. The noun in this case is `token`; there is no other identifier. 

`ipv6-address-token`

**Comment by slyon** [comment]
> You could drop the `id` portion of the schema key. The noun in this case is `token`; there is no other identifier.
> 
> `ipv6-address-token`

Yeah... I was thinking about this as well... but this token is basically the "IPv6 interface identifier", so I wanted to add the `-id` part.
Maybe `ipv6-address-interface-id` would be a good name?

**Comment by cluonbeam** [comment]
That's actually not what a token is. A token allows the administrator to set the host portion of the address to a static value even though the rest of the configuration is some form of autoconfiguration. Also, interface ID is **very ** ambiguous with other configuration elements. The only truly accurate and unambiguous name for this is `ipv6-address-token`

File: src/parse.c [hunk 1]
```diff
@@ -1735,7 +1746,7 @@ static const mapping_entry_handler nameservers_handlers[] = {
     {"route-metric", YAML_SCALAR_NODE, handle_netdef_guint, NULL, netdef_offset(overrides.metric)},         \
     {"send-hostname", YAML_SCALAR_NODE, handle_netdef_bool, NULL, netdef_offset(overrides.send_hostname)},  \
     {"use-dns", YAML_SCALAR_NODE, handle_netdef_bool, NULL, netdef_offset(overrides.use_dns)},              \
-    {"use-domains", YAML_SCALAR_NODE, handle_netdef_str, NULL, netdef_offset(overrides.use_domains)},      \
+    {"use-domains", YAML_SCALAR_NODE, handle_netdef_str, NULL, netdef_offset(overrides.use_domains)},       \
```
Comment (vorlonofportland) [inline src/parse.c:1749]: correct change, but unrelated?

File: src/parse.c [hunk 2]
```diff
@@ -1780,11 +1792,11 @@ static const mapping_entry_handler dhcp6_overrides_handlers[] = {
...
-#define PHYSICAL_LINK_HANDLERS                                                           \
-    {"match", YAML_MAPPING_NODE, handle_match},                                          \
-    {"set-name", YAML_SCALAR_NODE, handle_netdef_str, NULL, netdef_offset(set_name)},    \
-    {"wakeonlan", YAML_SCALAR_NODE, handle_netdef_bool, NULL, netdef_offset(wake_on_lan)}, \
-    {"wakeonwlan", YAML_SEQUENCE_NODE, handle_wowlan, NULL, netdef_offset(wowlan)},       \
+#define PHYSICAL_LINK_HANDLERS                                                                \
+    {"match", YAML_MAPPING_NODE, handle_match},                                               \
+    {"set-name", YAML_SCALAR_NODE, handle_netdef_str, NULL, netdef_offset(set_name)},         \
+    {"wakeonlan", YAML_SCALAR_NODE, handle_netdef_bool, NULL, netdef_offset(wake_on_lan)},    \
+    {"wakeonwlan", YAML_SEQUENCE_NODE, handle_wowlan, NULL, netdef_offset(wowlan)},           \
```
Comment (vorlonofportland) [inline src/parse.c:1799]: more unrelated whitespace changes

File: src/validation.c [hunk 3]
```diff
@@ -195,6 +195,9 @@ validate_netdef_grammar(NetplanNetDefinition* nd, yaml_node_t* node, GError** er
             goto netdef_grammar_error;
     }
 
+    if (nd->ip6_addr_gen_mode > NETPLAN_ADDRGEN_EUI64 && nd->ip6_addr_gen_token)
+        return yaml_error(node, error, "%s: ipv6-address-generation and ipv6-address-token-id are mutually exclusive", nd->id);
+
```
Comment (vorlonofportland) [inline src/validation.c:200]: I'm uncomfortable with this being an inequality comparison. If other values are added to NetplanAddrGenMode in the future, what guarantees these checks are correct?

File: doc/netplan.md [hunk 4]
```diff
@@ -272,6 +272,12 @@ Virtual devices
     Stateless Address Autoconfiguration (only supported with `NetworkManager`
     backend). Possible values are ``eui64`` or ``stable-privacy``.
 
+``ipv6-address-token-id`` (scalar) – since **0.100**
+
+:   Define an IPv6 address token for creating a static interface identifier for
+    IPv6 Stateless Address Autoconfiguration. This implies a slightly modified
+    ``eui64`` mode and is mutually exclusive with ``ipv6-address-generation``.
+
```
Comment (vorlonofportland) [inline doc/netplan.md:280]: I agree with the other PR comments: "token id" is redundant.

I also think that "token" itself is unclear; and at least in networkd, token is used to mean something other than just the id.

https://tools.ietf.org/html/rfc4862 mentions in Appendix B that they /used to/ use the term "interface token" but now say "interface identifier".

At https://developer.gnome.org/NetworkManager/stable/settings-ipv6.html, NetworkManager references draft-chown-6man-tokenised-ipv6-identifiers-02 for its definition of "token". But this also refers to "tokenized interface identifier" - so "interface identifier" is used consistently.

Therefore I recommend "ipv6-interface-identifier" as the key for this.

Comment (cluonbeam) [reply doc/netplan.md:280]: The term used in iproute2, NetworkManager, systemd, etc. is "token." The token is **not** an interface identifier; labeling it as such is extremely misleading. The accurate key for this is, as stated, `ipv6-address-token`.

Comment (vorlonofportland) [reply doc/netplan.md:280]: I agree it's important that this be discoverable for people who are looking for "token", given that this term is used throughout iproute2, NetworkManager, and networkd as you mention (though, not with a consistent meaning).

Given that the term 'token' is deprecated in the RFCs, however, I don't like `ipv6-token-id` as this propagates the idea of a 'token' as something that new users are expected to understand.

My suggestions are either to use the `ipv6-interface-identifier` and reference 'token' in the documentation; or use `ipv6-address-token`.

File: src/parse.c (hunk 1 above)
Comment (slyon) [reply src/parse.c]: Yeah... unrelated. I removed them and will prepare another PR for cleaning this up.

File: src/parse.c (hunk 2 above)
Comment (slyon) [reply src/parse.c]: I removed them and will prepare another PR for cleaning this up.

File: src/validation.c (hunk 3 above)
Comment (slyon) [reply src/validation.c:200]: Good point! I switched to a `!= NETPLAN_ADDRGEN_DEFAULT` comparison, which makes it more explicit to fail if `ipv6-address-generation` is set in any way.

File: doc/netplan.md (hunk 4 above)
Comment (slyon) [reply doc/netplan.md:280]: Thank you very much @cluonbeam and @vorlonofportland for your input!
I'm going with `ipv6-address-token` then, to keep it in line with the related `ipv6-address-generation` setting.

**Review by sil2100 — APPROVED** [review]
Ok, this looks good. I see all Steve's review comments have also been addressed, so I think this is good to go.
